File: Important Techniques/findingMidinLL.py
class Node:
    def __init__(self,value) -> None:
        self.data = value
        self.next = None


head = None

def findMid(head):
    slow = head
    fast = head
    if(not head):
        return
    elif head.next is None or head.next.next is None:
        return head.data
    else:
        # fast.next is not None => for odd number of nodes in LL
        # fast.next.next is not None => for even number of nodes in LL
        while(fast.next is not None and fast.next.next is not None):
            slow = slow.next
            fast = fast.next.next
        return slow.data
head = Node(1)
head.next = Node(16)
head.next.next = Node(15)
head.next.next.next = Node(12)
# head.next.next.next.next = Node(22)

if(findMid(head)):
    print(findMid(head))
else:
    print("No mid")
            

File: Important Techniques/findingMidinLL.c
#include<stdio.h>
#include<stdlib.h>

struct Node
{
    int data;
    struct Node *link;  
};

typedef struct Node Node;
Node *head = NULL;

Node *newNode(int value){
    Node *temp = (Node*)malloc(sizeof(Node));
    if (temp == NULL){
        printf("No memory allocated\n");
    }
    temp->data = value;
    temp->link = NULL;
    return temp;
}

int insertAtBeginSingle(int value){
    if(head == NULL){
        head = newNode(value);
    }
    else{
        Node *temp = newNode(value);
        temp->link = head;
        head = temp;
    }
}

int findMid(){
    Node* slow = head;
    Node* fast = head;

    if(!head){
        return -1;
    }
    else if ((head->link == NULL) || (head->link->link == NULL)){
        return head->data;
    }
    else{
        while ((fast->link != NULL) && (fast->link->link != NULL))
        {
            slow = slow->link;
            fast = fast->link->link;
        }
        return slow->data;  
    }   
}

int main(){
    insertAtBeginSingle(12);
    insertAtBeginSingle(17);
    insertAtBeginSingle(15);
    insertAtBeginSingle(55);
    insertAtBeginSingle(35);
    insertAtBeginSingle(25);
    printf("%d",findMid());
}